borgmatic/commands/borgmatic.py

```diff
@@ -84,26 +85,33 @@ def main():  # pragma: no cover
 
         for config_filename in config_filenames:
             config = validate.parse_configuration(config_filename, validate.schema_filename())
-            (location, storage, retention, consistency) = (
+            (location, storage, retention, consistency, hooks) = (
                 config.get(section_name, {})
-                for section_name in ('location', 'storage', 'retention', 'consistency')
+                for section_name in ('location', 'storage', 'retention', 'consistency', 'hooks')
             )
             remote_path = location.get('remote_path')
 
-            create.initialize(storage)
+            try:
+                create.initialize(storage)
+                hook.execute_hook(hooks.get('before_backup'))
 
-            for repository in location['repositories']:
-                if args.prune:
-                    prune.prune_archives(args.verbosity, repository, retention, remote_path=remote_path)
-                if args.create:
-                    create.create_archive(
-                        args.verbosity,
-                        repository,
-                        location,
-                        storage,
-                    )
-                if args.check:
-                    check.check_archives(args.verbosity, repository, consistency, remote_path=remote_path)
+                for repository in location['repositories']:
+                    if args.prune:
+                        prune.prune_archives(args.verbosity, repository, retention, remote_path=remote_path)
+                    if args.create:
+                        create.create_archive(
+                            args.verbosity,
+                            repository,
+                            location,
+                            storage,
+                        )
+                    if args.check:
+                        check.check_archives(args.verbosity, repository, consistency, remote_path=remote_path)
+
+                hook.execute_hook(hooks.get('after_backup'))
+            except (OSError, CalledProcessError):
```

Comment:
Given that you're re-raising the exception, you *could* `except:` all errors here. It kind of depends on what you want the contract to be for "on_error" hooks.. Should they trigger on all error types or just these particular ones?

Reply:
Catching all errors with ```except:``` would imply that an unhandled error would be raised as well. Since the outer exception handler can only deal with ```ValueError, OSError, CalledProcessError```, unhandled exceptions would still be fatal. Hence, it probably makes sense to keep the inner ```except``` consistent with the outer.

Reply:
Yup, makes sense. Just didn't know if you wanted fatal exceptions to call the "on_error" hook as well before.. being fatal.